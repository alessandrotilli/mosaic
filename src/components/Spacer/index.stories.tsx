import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import getDocsPage from "../../utils/stories";

import Spacer, { DATA_CY_DEFAULT } from ".";

const meta = {
  title: "Display/Spacer",
  component: Spacer,
  decorators: [
    (Story, { args: { direction } }) => (
      <div id="flex-content" style={{ display: "flex", flexDirection: direction === "vertical" ? "column" : "row" }}>
        <div id="first-content">First Content</div>
        {Story()}
        <div id="second-content">Second Content</div>
      </div>
    ),
  ],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: { label: "HTML Div" },
        component: "Spacer",
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
      }),
    },
  },
} satisfies Meta<typeof Spacer>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {},
};

export const InvalidSpacing: Story = {
  args: {
    ...Primary.args,
    level: -999,
  },
};

export const HorizontalSpacing: Story = {
  args: {
    ...Primary.args,
    direction: "horizontal",
  },
};

export const VerticalSpacing: Story = {
  args: {
    ...Primary.args,
    direction: "vertical",
  },
};
