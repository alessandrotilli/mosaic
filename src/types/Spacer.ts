import { IBase } from "./Base";

export enum SpacerDirection {
  horizontal = "horizontal",
  vertical = "vertical",
}

export interface SpacerType extends IBase {
  direction?: SpacerDirection;
  level?: number;
}
