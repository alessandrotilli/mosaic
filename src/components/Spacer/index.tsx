import React, { CSSProperties, FC, memo, useMemo } from "react";
import { useTheme } from "@mui/material";

import { ISpacer } from "../../types/Spacer";

export const DATA_CY_DEFAULT = "spacer";

const Spacer: FC<ISpacer> = ({ dataCy = DATA_CY_DEFAULT, direction = "horizontal", level: externalLevel = 1 }) => {
  const { spacing } = useTheme();

  const level = useMemo(() => (externalLevel < 0 || !externalLevel ? 1 : externalLevel), [externalLevel]);

  const height = useMemo(() => (direction === "horizontal" ? 0 : spacing(level)), [direction, level, spacing]);

  const width = useMemo(() => (direction === "horizontal" ? spacing(level) : 0), [direction, level, spacing]);

  const style = useMemo(
    (): CSSProperties => ({
      display: "block",
      height,
      width,
    }),
    [height, width]
  );

  return <div data-cy={dataCy} style={style} />;
};

export const MemoizedSpacer = memo(Spacer);

export default MemoizedSpacer;
