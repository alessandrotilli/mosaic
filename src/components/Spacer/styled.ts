import { Box as MUIBox, styled } from "@material-ui/core";
import { SpacerDirection } from "../../types/Spacer";

// TODO: this (props: any) here should be removed
export const StyledMUIBox = styled(MUIBox)((props: any) => {
  const { direction, level } = props;
  const size = level * 8;
  return {
    display: "block",
    height: direction === SpacerDirection.horizontal ? 0 : size,
    width: direction === SpacerDirection.horizontal ? size : 0,
  };
});
