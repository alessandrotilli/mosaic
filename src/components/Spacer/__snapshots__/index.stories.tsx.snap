// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display/Spacer HorizontalSpacing smoke-test 1`] = `
<div id="flex-content"
     style="display: flex; flex-direction: row;"
>
  <div id="first-content">
    First Content
  </div>
  <div data-cy="spacer"
       style="display: block; height: 0px; width: 8px;"
  >
  </div>
  <div id="second-content">
    Second Content
  </div>
</div>
`;

exports[`Display/Spacer InvalidSpacing smoke-test 1`] = `
<div id="flex-content"
     style="display: flex; flex-direction: row;"
>
  <div id="first-content">
    First Content
  </div>
  <div data-cy="spacer"
       style="display: block; height: 0px; width: 8px;"
  >
  </div>
  <div id="second-content">
    Second Content
  </div>
</div>
`;

exports[`Display/Spacer Primary smoke-test 1`] = `
<div id="flex-content"
     style="display: flex; flex-direction: row;"
>
  <div id="first-content">
    First Content
  </div>
  <div data-cy="spacer"
       style="display: block; height: 0px; width: 8px;"
  >
  </div>
  <div id="second-content">
    Second Content
  </div>
</div>
`;

exports[`Display/Spacer VerticalSpacing smoke-test 1`] = `
<div id="flex-content"
     style="display: flex; flex-direction: column;"
>
  <div id="first-content">
    First Content
  </div>
  <div data-cy="spacer"
       style="display: block; height: 8px; width: 0px;"
  >
  </div>
  <div id="second-content">
    Second Content
  </div>
</div>
`;
